pkg/controller/complianceremediation/complianceremediation_controller.go

```diff
@@ -0,0 +1,380 @@
+	if err != nil {
+		return &ReconcileComplianceRemediation{}
+	}
+
+	return &ReconcileComplianceRemediation{client: mgr.GetClient(),
+		mcClient: mcClient,
+		scheme:   mgr.GetScheme()}
+}
+
+// add adds a new Controller to mgr with r as the reconcile.Reconciler
+func add(mgr manager.Manager, r reconcile.Reconciler) error {
+	// Create a new controller
+	c, err := controller.New("complianceremediation-controller", mgr, controller.Options{Reconciler: r})
+	if err != nil {
+		return err
+	}
+
+	// Watch for changes to primary resource ComplianceRemediation
+	err = c.Watch(&source.Kind{Type: &complianceoperatorv1alpha1.ComplianceRemediation{}}, &handler.EnqueueRequestForObject{})
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// blank assignment to verify that ReconcileComplianceRemediation implements reconcile.Reconciler
+var _ reconcile.Reconciler = &ReconcileComplianceRemediation{}
+
+// ReconcileComplianceRemediation reconciles a ComplianceRemediation object
+type ReconcileComplianceRemediation struct {
+	// This client, initialized using mgr.Client() above, is a split client
+	// that reads objects from the cache and writes to the apiserver
+	client   client.Client
+	mcClient *mcfgClient.MachineconfigurationV1Client
+	scheme   *runtime.Scheme
+}
+
+// Reconcile reads that state of the cluster for a ComplianceRemediation object and makes changes based on the state read
+// and what is in the ComplianceRemediation.Spec
+// Note:
+// The Controller will requeue the Request to be processed again if the returned error is non-nil or
+// Result.Requeue is true, otherwise upon completion it will remove the work from the queue.
+func (r *ReconcileComplianceRemediation) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	var err error
+
+	reqLogger := log.WithValues("Request.Namespace", request.Namespace, "Request.Name", request.Name)
+	reqLogger.Info("Reconciling ComplianceRemediation")
+
+	// Fetch the ComplianceRemediation instance
+	remediationInstance := &complianceoperatorv1alpha1.ComplianceRemediation{}
+	err = r.client.Get(context.TODO(), request.NamespacedName, remediationInstance)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			// Request object not found, could have been deleted after reconcile request.
+			// Owned objects are automatically garbage collected. For additional cleanup logic use finalizers.
+			// Return and don't requeue
+			return reconcile.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		return reconcile.Result{}, err
+	}
+
+	if remediationInstance.Spec.Apply == false {
+		reqLogger.Info("Won't reconcile remediation which is not selected", "remediation", request.NamespacedName)
+		return reconcile.Result{}, nil
+	}
+
+	if remediationInstance.Status.ApplicationState == complianceoperatorv1alpha1.RemediationApplied {
+		reqLogger.Info("Won't reconcile remediation which is already applied", "remediation", request.NamespacedName)
+		return reconcile.Result{}, nil
+	}
```

Comment:
I don't think this if statement is necessary. we should enforce the state. If it's in "applied state" it should be safe to try to create each time the remediation. If the MC already exists, that's not a problem, just means we'll fail the create call. Ideally if the remediation is edited, we should update the MC.

Reply:
I tend to agree it would be more inline with the general style of just reconciling the state. I'll run some tests.

Reply:
did you get a chance to try this?

Reply:
Yes and not only you were right, but the logic in the remediations controller was stupid in the sense that it did not un-apply remediations that were toggled from spec.apply=true to spec.apply=false. I fixed this and also added code that, if the resulting MC is empty, removes the remediation MC. So the admin could now both select and de-select remediations and after all remediations are de-selected, would get into the original state.